eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -542,89 +549,187 @@ instance Semigroup ScriptResult where
-      [PV2.Data]
-      ProtVer
-  deriving (Eq, Generic, NoThunks)
+data PlutusData :: Language -> Type where
+  DataV1 :: {unDataV1 :: [PCD.Data]} -> PlutusData 'PlutusV1
+  DataV2 :: {unDataV2 :: [PCD.Data]} -> PlutusData 'PlutusV2
+
+deriving instance Eq (PlutusData l)
+
+mkPlutusData :: SLanguage l -> [PCD.Data] -> PlutusData l
+mkPlutusData = \case
+  SPlutusV1 -> DataV1
+  SPlutusV2 -> DataV2
+
+instance forall (l :: Language). (Typeable l, IsLanguage l) => ToCBOR (PlutusData l) where
+  toCBOR = \case
+    DataV1 dl -> toCBOR PlutusV1 <> encodeList toCBOR dl
+    DataV2 dl -> toCBOR PlutusV2 <> encodeList toCBOR dl
+
+instance forall (l :: Language). (Typeable l, IsLanguage l) => FromCBOR (PlutusData l) where
+  fromCBOR = do
+    _ <- fromCBOR @Language
+    mkPlutusData isLanguage <$> decodeList fromCBOR
+
+data PlutusDebug (l :: Language) where
+  PlutusDebug ::
+    { pdCostModel :: CostModel,
+      pdExUnits :: ExUnits,
+      pdSBS :: SBS.ShortByteString,
+      pdPlutusData :: PlutusData l,
+      pdProtVer :: ProtVer
+    } ->
+    PlutusDebug l
 
 -- | There is dummy Show instance for PlutusDebug intentionally, because it is too
 -- expensive and it will be too tempting to use it incorrectly. If needed for
 -- testing use 'StandaloneDeriving', otherwise define an efficient way to display
 -- this info.
-instance Show PlutusDebug where
+instance Show (PlutusDebug l) where
   show _ = "PlutusDebug Omitted"
 
-data PlutusError = PlutusErrorV1 PV1.EvaluationError | PlutusErrorV2 PV2.EvaluationError
+deriving instance Eq (PlutusDebug l)
+
+deriving instance Generic (PlutusDebug l)
+
+mkPlutusDebug ::
+  forall (l :: Language).
+  SLanguage l ->
+  CostModel ->
+  ExUnits ->
+  SBS.ShortByteString ->
+  [PCD.Data] ->
+  ProtVer ->
+  PlutusDebug l
+mkPlutusDebug sl costModel exUnits sbs pData protVer = case sl of
+  SPlutusV1 -> PlutusDebug costModel exUnits sbs (DataV1 pData) protVer
+  SPlutusV2 -> PlutusDebug costModel exUnits sbs (DataV2 pData) protVer
+
+plutusDebugLang ::
+  forall (l :: Language).
+  (IsLanguage l) =>
+  PlutusDebug l ->
+  (SLanguage l, PlutusDebug l)
+plutusDebugLang p = (isLanguage, p)
+
+instance
+  forall (l :: Language).
+  (Typeable l, IsLanguage l) =>
+  ToCBOR (PlutusDebug l)
+  where
+  toCBOR (PlutusDebug costModel exUnits sbs pData protVer) = case pData of
+    DataV1 _ ->
+      encode $ Sum PlutusDebug 0 !> E encodeCostModel costModel !> To exUnits !> To sbs !> To pData !> To protVer
+    DataV2 _ ->
+      encode $ Sum PlutusDebug 1 !> E encodeCostModel costModel !> To exUnits !> To sbs !> To pData !> To protVer
+
+instance
+  forall (l :: Language).
+  (Typeable l, IsLanguage l) =>
+  FromCBOR (PlutusDebug l)
+  where
+  fromCBOR = do
+    language <- fromCBOR @Language
+    costModel <- decodeCostModel language
+    exUnits <- fromCBOR
+    sbs <- fromCBOR
+    pData <- fromCBOR
+    protVer <- fromCBOR
+    pure $ mkPlutusDebug isLanguage costModel exUnits sbs pData protVer
```

Comment:
There is no check in this CBOR instance that the `isLanguage`. So you need a helper function:
```suggestion
    language <- fromCBOR @Language
    thisLanguage <- toSLanguage language
    costModel <- decodeCostModel language
    exUnits <- fromCBOR
    sbs <- fromCBOR
    pData <- fromCBOR
    protVer <- fromCBOR
    pure $ PlutusDebug thisLanguage costModel exUnits sbs pData protVer
```
This would is the helper function that can go to `Cardano.Ledger.Language`:

```haskell
toSLanguage :: (IsLanguage l, MonadFail m) => Language -> m (SLanguage l)
toSLanguage lang
  | fromSLanguage thisLanguage == lang = pure thisLanguage
  | otherwise = fail $ "Plutus language mismatch. Expected "  ++ show slang ++ ", but got: " ++ show lang
  where
   thisLanguage = isLanguage
```

Reply:
How about
```haskell
instance
  forall (l :: Language).
  (Typeable l, IsLanguage l) =>
  FromCBOR (PlutusDebugLang l)
  where
  fromCBOR = do
    slang <- fromCBOR @(SLanguage l)
    costModel <- decodeCostModel (fromSLanguage slang)
    exUnits <- fromCBOR
    sbs <- fromCBOR
    pData <- fromCBOR
    protVer <- fromCBOR
    pure $ PlutusDebugLang slang costModel exUnits sbs pData protVer
```
where `ToCBOR` and `FromCBOR` for `SLanguage` is the same as that for `Langauge`?

Reply:
Sure that would work, but the `FromCBOR (SLanguage l)` will have to use some functionality like: `toSLanguage`

Side note type application `@(SLanguage l)` would not be necessary, ghc will be able to infer it